pkg/controller/admin/realms.go

```diff
@@ -199,3 +201,130 @@ func (c *Controller) renderEditRealm(ctx context.Context, w http.ResponseWriter,
 	m["supportsPerRealmSigning"] = c.db.SupportsPerRealmSigning()
 	c.h.RenderHTML(w, "admin/realms/edit", m)
 }
+
+func (c *Controller) HandleRealmsJoin() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		vars := mux.Vars(r)
+
+		session := controller.SessionFromContext(ctx)
+		if session == nil {
+			controller.MissingSession(w, r, c.h)
+			return
+		}
+		flash := controller.Flash(session)
+
+		user := controller.UserFromContext(ctx)
+		if user == nil {
+			controller.MissingUser(w, r, c.h)
+			return
+		}
+
+		realm, err := c.db.FindRealm(vars["id"])
+		if err != nil {
+			controller.InternalError(w, r, c.h, err)
+			return
+		}
+
+		user.Realms = append(user.Realms, realm)
+		user.AdminRealms = append(user.AdminRealms, realm)
+
+		// Do the membership update and audit entry in a transaction because we need
+		// both to succeed to continue.
+		if err := c.db.RawDB().Transaction(func(tx *gorm.DB) error {
```

Comment:
I'm really not a fan of this, but I can't seem to find a better way. I thought about using callbacks, but there's far too many fields to set on the auditable resource.

I'm was thinking it might be better to allow auditable resource to have a *AuditEntry field that can be nil. If it's populated, have the callback create the entry. I just... couldn't get it to work with gorm.